File: rd/tag/2016.1.21/Server/ServerCode/hawkjava/HawkClient/HawkUtil/HawkProtocol.cpp
#include "HawkProtocol.h"
#include "HawkOSOperator.h"
#include "HawkLoggerManager.h"
#include "HawkStringUtil.h"
#include "HawkProtocolManager.h"

namespace Hawk
{
	HawkProtocol::HawkProtocol(ProtoType iType, OctetsStream* pOctets) : m_iType(iType)
	{
		if (pOctets != 0) {
			Encode(*pOctets);
		}
	}

	HawkProtocol::~HawkProtocol()
	{
	}

	ProtoType HawkProtocol::GetType() const
	{
		return m_iType;
	}	

	ProtoSize HawkProtocol::GetSize() const
	{
		return m_iSize;
	}

	ProtoCrc HawkProtocol::GetCrc() const
	{
		return m_iCrc;
	}

	Bool HawkProtocol::GetOctets(OctetsStream*& pOS)
	{
		pOS = &m_sDecode;
		return true;
	}

	Bool HawkProtocol::Encode(const OctetsStream& rhsOS)
	{
		try
		{				
			m_iSize = rhsOS.Size();
			m_iCrc  = HawkOSOperator::CalcCrc((UChar*)rhsOS.Begin(), m_iSize);
			
			m_sDecode.Clear();
			P_ProtocolManager->WriteProtocolHeader(m_sDecode, m_iType, m_iSize, m_iCrc);
			m_sDecode.Push(rhsOS.Begin(),m_iSize);
		}
		catch (HawkException& rhsExcep)
		{
			HawkFmtError(rhsExcep.GetMsg().c_str());
			return false;
		}

		return true;
	}

	Bool HawkProtocol::Decode(HawkOctetsStream& rhsOS)
	{
		//检测协议完整性
		if (!P_ProtocolManager->CheckDecodeProtocol(rhsOS))
			return false;
		
		m_iSize = 0;
		m_iCrc  = 0;
		m_sDecode.Clear();

		ProtoType iType    = 0;
		ProtoCrc  iCrc     = 0;
		try
		{		
			//读协议头
			rhsOS >> HawkOctetsStream::TransBegin;

			P_ProtocolManager->ReadProtocolHeader(rhsOS, iType, m_iSize, m_iCrc);
			
			if ((ProtoSize)rhsOS.AvailableSize() >= m_iSize)
			{
				//读协议数据
				if (m_iSize)
				{
					m_sDecode.Reserve(m_iSize);
					rhsOS.Pop((void*)m_sDecode.Begin(), m_iSize);
					m_sDecode.Resize(m_iSize);
					iCrc = HawkOSOperator::CalcCrc((UChar*)m_sDecode.Begin(),m_iSize);
				}

				//读取完全,提交
				rhsOS >> HawkOctetsStream::TransCommit;
			}
			else
			{
				//数据包不完全,回滚
				rhsOS >> HawkOctetsStream::TransRollback;
				return false;
			}
		}
		catch (HawkException& rhsExcep)
		{
			rhsOS >> HawkOctetsStream::TransRollback;
			HawkFmtError(rhsExcep.GetMsg().c_str());
			return false;
		}

		//数据校验
		if(m_iType != iType)
		{
			HawkFmtError("Protocol Type Inconformity, Type: %u, Crc: %u", iType, m_iCrc);
			T_Exception("Protocol Type Inconformity.");
			return false;
		}
		else if(m_iCrc != iCrc)
		{
			HawkFmtError("Protocol Crc Inconformity, Type: %u, Crc: %u", iType, m_iCrc);
			T_Exception("Protocol Crc Inconformity.");
			return false;
		}

		return true;
	}
}

File: rd/tag/2016.1.21/Server/ServerCode/hawkjava/Tools/websocket/script/hawk/HawkProtocol.js
/**
 * Created by hawk on 2014/9/13.
 */

// 构造
function HawkProtocol(type, msgBuilder) {
    if (type && msgBuilder) {
        this.type = type;
		try {
			this.msgBuffer = msgBuilder.encode();
		} catch (e) {
			console.log("builder encode failed: " + type + ", exception: " + e);
		}		
        this.size = this.msgBuffer.remaining();
        this.reserve = 0;
        this.crc = HawkProtocolManager.calcCrc(this.msgBuffer);
    } else {
        this.type = 0;
        this.size = 0;
        this.reserve = 0;
        this.crc = 0;
        this.msgBuffer = null;
    }
}

HawkProtocol.prototype.getType = function () {
    return this.type;
}

HawkProtocol.prototype.getSize = function () {
    return this.size;
}

HawkProtocol.prototype.getReserve = function () {
    return this.reserve;
}

HawkProtocol.prototype.getCrc = function () {
    return this.crc;
}

HawkProtocol.prototype.getMsgBuffer = function () {
    return this.msgBuffer;
}

// 协议编码, 返回ByteBuffer
HawkProtocol.prototype.encode = function() {
    try {
        var byteBuffer = new ByteBuffer(1024);
        byteBuffer.writeInt32(this.type);
        byteBuffer.writeInt32(this.size);
        byteBuffer.writeInt32(this.reserve);
        byteBuffer.writeInt32(this.crc);
        byteBuffer.append(this.msgBuffer);
        byteBuffer.flip();
        return byteBuffer;
    } catch (e) {
		console.log("protocol encode failed: " + type + ", exception: " + e);
    }
    return null;
}

// 协议解码
HawkProtocol.prototype.decode = function(byteBuffer) {
    if (byteBuffer.remaining() < HawkProtocolManager.HEADER_SIZE) {
        return false;
    }
    byteBuffer.mark();
    try {
        this.type = byteBuffer.readInt32();
        this.size = byteBuffer.readInt32();
        this.reserve = byteBuffer.readInt32();
        this.crc = byteBuffer.readInt32();
        if (byteBuffer.remaining() >= this.size) {
            // 读协议数据
			this.msgBuffer = new ByteBuffer(this.size);
            if (this.size > 0) {
				byteBuffer.copyTo(this.msgBuffer, 0, byteBuffer.offset, byteBuffer.offset + this.size);			
            }
        } else {
            // 数据包不完全,回滚
            byteBuffer.reset();
            return false;
        }
    } catch (e) {
        // 数据包不完全,回滚
        byteBuffer.reset();
        console.log("exception: " + e);
        return false;
    }

    // crc校验
    var calcCrc = HawkProtocolManager.calcCrc(this.msgBuffer);
    if (calcCrc != this.crc) {
        return false;
    }
    return true;
}
